boa/interpret.py

```diff
@@ -286,6 +301,73 @@ def _compile():
     return _disk_cache.caching_lookup(cache_key, _compile)
 
 
+def _create_compiler_data(
+    path: Path, resolved_path: Path, source_code: str | bytes, settings: Settings
+) -> CompilerData:
+    try:
+        return _create_archive_compiler_data(source_code, settings)
+    except NotZipInput:
+        pass
+
+    if isinstance(source_code, bytes):
+        source_code = source_code.decode()
+
+    global _search_path
+    file_input = FileInput(
+        contents=source_code, source_id=-1, path=path, resolved_path=resolved_path
+    )
+    input_bundle = FilesystemInputBundle(get_search_paths(_search_path))
+    return CompilerData(file_input, input_bundle, settings)
+
+
+def _create_archive_compiler_data(
+    zip_contents: str | bytes, settings: Settings
+) -> CompilerData:
+    with _open_zip(zip_contents) as archive:
+        # read the whole zip into memory so it can be serialized to the cache
+        files = {name: archive.read(name).decode() for name in archive.namelist()}
+
+    targets = files["MANIFEST/compilation_targets"].splitlines()
+    if len(targets) != 1:
+        raise BadArchive("Multiple compilation targets not supported!")
+
+    input_bundle = JSONInputBundle(
+        input_json={
+            PurePath(name): {"content": content} for name, content in files.items()
+        },
+        search_paths=[PurePath(p) for p in files["MANIFEST/searchpaths"].splitlines()],
+    )
+
+    main_path = PurePath(targets[0])
+    file = input_bundle.load_file(main_path)
+    assert isinstance(file, FileInput)  # help mypy
+    settings_json = json.loads(files["MANIFEST/settings.json"])
+    settings = merge_settings(
+        settings,
+        Settings.from_dict(settings_json),
+        lhs_source="command line",
+        rhs_source="archive settings",
+    )
+    integrity_sum = files["MANIFEST/integrity"].strip()
+    return CompilerData(file, input_bundle, settings, integrity_sum)
+
+
+def _open_zip(zip_contents):
+    if isinstance(zip_contents, str):
+        zip_contents = zip_contents.encode()
+    try:
+        buf = BytesIO(zip_contents)
+        return ZipFile(buf, mode="r")
+    except BadZipFile as e1:
+        try:
+            # don't validate base64 to allow for newlines
+            zip_contents = b64decode(zip_contents, validate=False)
+            buf = BytesIO(zip_contents)
+            return ZipFile(buf, mode="r")
+        except (BadZipFile, binascii.Error):
+            raise NotZipInput() from e1
+
```

Comment:
i think we should refactor the compiler so that we can return CompilerData

Reply:
i suppose we could keep it this way but with a note to delete after refactored in vyper

Reply:
that conversion of the input bundle is a bit awkward though